### ISSUE TITLE
[BUG] bunkrr.su and simpcity download doesn't work

### ISSUE THREAD

**Issue by wdos3** [Jules-WinnfieldX/CyberDropDownloader#536]
Version information: Python 3.9.12, Windows 11

As mentioned above, simpcity scrape doesn't work, and bunkrr seems like it's blocked:
- Simpcity simply returns an error message for the entire URL, possibly because it's simpcity.su instead of .st (though I am not sure whether that's the real reason, or the site has banned scraping, or traffic restrictions)
- bunkrr seems to also suffer from traffic restrictions, as videos can be manually downloaded through the website instead of a scraper. If it helps, the scraped video returns the error twice (for example, it repeated "Error: https://bunkrr.su/v/0he0shdwyaj7wrz8pxc1j_source-FsDxNZEs.mp4" twice instead of only once)

**Comment by Jules-WinnfieldX** [comment]
So here's the gist.

1. Simpcity has turned on captcha verification for login. Nobody should be surprised that this breaks Cyberdrop-DL. I have a way around it but it's going to take a long time to implement.
2. Bunkrr completely removed the section of the page where I was previously able to get download links from. It's going to take a hot minute before I am able to get a new methodology working.

**Comment by LuciferxR** [comment]
This can't happen, then we won't be able to download anything else.

**Comment by the444xg** [comment]
Saving the cookies on a file(or db) would help minimize those issues in the future. It's not a fix for the captcha but if the user has a session cookie already it won't be affected by those changes or eventual html or other simple protections. I know you said you don't wanna deal with cookies, but it's not that much work, at least for the forums.

On the Bunkrr issue, seems like they are using js to encode the server url. Also they lowered the limits and for some users they are showing recaptcha. I don't know how it was before but the player is using api to get the final video url.

**Comment by Jules-WinnfieldX** [comment]
> Saving the cookies on a file(or db) would help minimize those issues in the future. It's not a fix for the captcha but if the user has a session cookie already it won't be affected by those changes or eventual html or other simple protections. I know you said you don't wanna deal with cookies, but it's not that much work, at least for the forums.
> 
> On the Bunkrr issue, seems like they are using js to encode the server url. Also they lowered the limits and for some users they are showing recaptcha. I don't know how it was before but the player is using api to get the final video url.

Yea, I wanted to avoid messing with cookies as long as I could. Another project I have interest in has browser extraction baked in for the cookies. I'm going to go through with that. It's going to be part of a full rewrite though.

You're correct on the API for bunkr, I've been playing around with it this morning for the last hour, and it's full of 503 status'. I also need to figure out how this will apply to images / other filetypes. But I was just focused on video.

**Comment by the444xg** [comment]
> You're correct on the API for bunkr, I've been playing around with it this morning for the last hour, and it's full of 503 status'. I also need to figure out how this will apply to images / other filetypes. But I was just focused on video.

So far only saw videos in the errors, maybe images are not affected.



> Yea, I wanted to avoid messing with cookies as long as I could. Another project I have interest in has browser extraction baked in for the cookies. I'm going to go through with that. It's going to be part of a full rewrite though.

A temporary fix would be to use cookie extractor like https://raw.githubusercontent.com/instaloader/instaloader/master/docs/codesnippets/615_import_firefox_session.py

Or just a way to copy from browser and use it. Once you have the cookie it works fine, i believe i got one working for months.

After the login is fixed, something you can do to lower the number of requests normal users do is to implement the retry on failed/partial files. I feel like people would scrap everything again because it's hard to go one by one. No idea if this is too much work. For me if it was possible to add the thread id to the directory name, it would be much easier to keep track of those partial files with a script.

**Comment by github-userx** [comment]
@Jules-WinnfieldX looking forward to your rewrite and new implementations.

These site changes have once again been a great reminder to always locally backup files right away and never rely on streaming/cloudstorage sites as they can restrict content at any time or be gone completely from one day to another.

**Comment by github-userx** [comment]
What I do not understand is..bunkrr becomes basically useless for bulk upload/downloads if they don’t allow scrapers that get download 100 small videos in a row. It’s not like we’re downloading hundreds of videos manually,

**Comment by Jules-WinnfieldX** [comment]
> What I do not understand is..bunkrr becomes basically useless for bulk upload/downloads if they don’t allow scrapers that get download 100 small videos in a row. It’s not like we’re downloading hundreds of videos manually,

I won't trash talk bunkr. But it's more a game of cat and mouse. They have an issue with bot users, we want an easy way to download things. It just is what it is really.

**Comment by Jules-WinnfieldX** [comment]
> Note that the final API call to `https://api-v2.bunkrr.su/getFile` checks the `Referer` and uses a redirect to go to the appropriate file, so they need to be followed with `-L` in `curl`
> Now to see if pictures and albums change anything.

I have essentially this in my local test branch. Only video files have changed. The issue is Majority of the requests I make end up as HTTP 503 responses. Which is an issue on Bunkrr's side. I need to play around with it more. Another issue here is using download links such as this will entirely break the current methodology for download tracking/history. That to me is honestly the larger issue than switching everything to the API off the cuff. 

I don't know really how I'm going to be able to go about switching that. The current methodology uses the download links url path as the primary key (boiling down for simplicity). For those that don't know what that is, for this issue, the URL path would be `/Jules-WinnfieldX/CyberDropDownloader/issues/536`

On an API download link for bunkr videos right now, it'd be `/getFile`. The path doesn't include any of the queries, and as a second issue beyond that, the token query should never be considered static, it'll always change. So current history form, a single downloaded video will mark any potentially download as already complete.

**Comment by Jules-WinnfieldX** [comment]
If it was just swapping over to the API I'd have had it done this morning, 6 hours ago. But I need to try and wrap my head around how to make this work with the current state of the program. 

After I get it working (assuming I can) I'll be focusing on the re-write, which will involve fixing SC using cookie capture. It'll also mean I need to create a better DB for history tracking, while also trying to maintain backwards compatibility with the old history, or at the very least creating some method to convert it over on initial run.

**Comment by Jules-WinnfieldX** [comment]
> > The issue is Majority of the requests I make end up as HTTP 503 responses
> 
> Ouch, I haven't played with it enough to see these yet.
> 
> > On an API download link for bunkr videos right now, it'd be `/getFile`. The path doesn't include any of the queries, and as a second issue beyond that, the token query should never be considered static, it'll always change. So current history form, a single downloaded video will mark any potentially download as already complete.
> 
> The request to `/getFile?file_name=Dramatic-Look-%5By8Kyi0WNg40%5D-psx6dJa6.webm&tkn=blahblah` redirects to `https://media-files11.bunkr.ru/Dramatic-Look-[y8Kyi0WNg40]-psx6dJa6.webm`, so if the crawler code returns the redirect URL as opposed to the `/getFile` url, could this still be functional?

Ummm. Actually. If that's true, I can make a head request to the getFile api endpoint and just pull the response url. That would fix basically everything I'm talking about. I didn't think about that, or pay enough attention to see if it redirected. 

I'm technically sitting on a beach right now typing from my phone though, so I'll have to do that later tonight and push out an update if I can get everything to work.

**Comment by Jules-WinnfieldX** [comment]
It could be a *lot* cleaner, but 4.2.174 is going up now (when the bug strikes I suppose).

**Comment by roweger** [comment]
> It could be a _lot_ cleaner, but 4.2.174 is going up now (when the bug strikes I suppose).

> Edit: this should fix bunkr

Yep currently getting "error" from bunkr as well

**Comment by Jules-WinnfieldX** [comment]
Alrighty. Back to playing around with ratelimits...

**Comment by Jules-WinnfieldX** [comment]
> and those tended to be the ones that give 429s

Same tokens should be expected for some, as there's no caching it, would be pointless if your theory is correct on per second. That being said though, I just ran through scraping around 250 videos and I haven't gotten a single 429. 

You may need to provide me a url list to try.

**Comment by Jules-WinnfieldX** [comment]
I am getting timeout errors though for getToken. Trying to deal with that.

**Comment by Jules-WinnfieldX** [comment]
this get token method is making me mad. It's so inconsistent.

**Comment by Jules-WinnfieldX** [comment]
Pushing 4.2.175

The 404 is happening because bunkr is redirecting to a c*.bunkrr domain instead of a media-files*.bunkrr domain. It's a bunkrr issue essentially. 

This version is going to be a **lot** slower for bunkrr scrapes. I've limited the getToken and head requests to 1 req/s

**Comment by ClonedBoy** [comment]
Hi, thanks for this great program, has saved me multiple hours!

I am getting Errored scraping messages with SimpCity when using Brave (maybe this happens with other Chromium-based browsers?). I have tried the xf_session cookie value alone, with ' and ". Also tried different cookie values (logged out and logged back in multiple times).

When the value started with -, I got a message stating that the argument was invalid. 

When the value has - or _ in the string, the programs seems to not capture the correct cookie value, thus the error scraping message shows.

But when using Firefox, the cookie value generated was alphanumeric. I was able to get it working with the xf_session value from Firefox and not from Brave. The times I tried a new cookie in Brave, I didn't get it to be only alphanumeric. Maybe luck?
